apps/cyberstorm-nextjs/utils/auth.ts

```diff
@@ -0,0 +1,145 @@
+import { NextResponse } from "next/server";
+import type { NextRequest } from "next/server";
+import hkdf from "@panva/hkdf";
+import { EncryptJWT, jwtDecrypt } from "jose";
+import { v4 as uuid } from "uuid";
+
+interface UserStatusType {
+  username: string;
+  capabilities: string[];
+  subscription: { expires: string | null };
+  rated_packages: string[];
+  teams: string[];
+}
+
+function isUserStatusType(data: any): boolean {
+  if (!data.username) return false;
+  if (!data.capabilities) return false;
+  if (
+    typeof data.subscription.expires !== "string" &&
+    data.subscription.expires !== null
+  )
+    return false;
+  if (!data.subscription) return false;
+  if (!data.rated_packages) return false;
+  if (!data.teams) return false;
+  return true;
+}
+
+const now = () => (Date.now() / 1000) | 0;
+
+async function getDerivedEncryptionKey(secret: string | Buffer) {
+  return await hkdf(
+    "sha256",
+    secret,
+    "",
+    "Cyberstorm Generated Encryption Key",
+    32
+  );
+}
+
+async function encode(token: { username: string; expires: number }) {
+  if (process.env.CYBERSTORM_AUTH_SECRET) {
+    const encryptionSecret = await getDerivedEncryptionKey(
+      process.env.CYBERSTORM_AUTH_SECRET
+    );
+    return await new EncryptJWT(token)
+      .setProtectedHeader({ alg: "dir", enc: "A256GCM" })
+      .setIssuedAt()
+      .setExpirationTime(Math.min(now() + 900, token.expires))
+      .setJti(uuid())
+      .encrypt(encryptionSecret);
+  } else {
+    throw new Error("CYBERSTORM_AUTH_SECRET is missing");
+  }
+}
+
+export async function decode(
+  token: string
+): Promise<Record<string, unknown> | null> {
```

Comment:
Is there a reason to manually defining the return type (sometimes there is)? Mentioning it because it doesn't match what the current implementation can actually return.

Reply:
In this case I think, I again forgot to remove it. But sometimes I've also wanted to add those there before hand, so that I get type errors when developing and fucking around with the code 🤷 Bad practices maybe? I guess

Reply:
I haven't really formed an educated opinion on the matter. My gut feeling is that it's not inherently a bad practice, it can be a nice tool for some situations. If it's obvious that a function returns a string, it might be a bit of clutter to spell it out. If it's not obvious and you want to be 100% that some subsequent changes won't "widen" the return type accidentally, then spelling it out makes sense. But as seen here, spelling is out means you must also ensure it stays up to date if the return type is "narrowed". Having the spelled out type not match the actual return value is obviously bad, unless it's to make the function to implement some interface defined elsewhere.